### PR TITLE
TINKERPOP-3056 Fixed bug in mid-traversal mergeE for onMatch

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-3056

Since the current traverser was only being updated onMatch when it was used as a start step doing mutations in a sideEffect() would update whatever the current traverser was. This code may have been leftover from a time when the merge steps operated more readily on the current traverser as an argument to the step.

VOTE +1

### CODE REVIEW

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/MergeEdge.feature [hunk 1]
```diff
@@ -841,3 +841,45 @@ Feature: Step - mergeE()
...
+      """
+      g.V().mergeE(xx1).
+              option(Merge.onMatch, __.sideEffect(__.property("weight", 0)).constant([:]))
+      """
+    When iterated to list
+    Then the result should have a count of 2
+    And the graph should return 2 for count of "g.V()"
+    And the graph should return 0 for count of "g.E().hasLabel(\"knows\").has(\"weight\",1)"
+    And the graph should return 1 for count of "g.E().hasLabel(\"knows\").has(\"weight\",0)"
+    And the graph should return 0 for count of "g.V().hasLabel(\"knows\").has(\"weight\")"
```
Comment (vkagamlyk) [inline gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/MergeEdge.feature:864]:
```suggestion
    And the graph should return 0 for count of "g.V().has(\"weight\")"
```

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/MergeEdge.feature [hunk 2]
```diff
@@ -841,3 +841,45 @@ Feature: Step - mergeE()
...
+      """
+      g.mergeE(xx1).
+          option(Merge.onMatch, __.sideEffect(__.property("weight", 0)).constant([:]))
+      """
+    When iterated to list
+    Then the result should have a count of 1
+    And the graph should return 2 for count of "g.V()"
+    And the graph should return 0 for count of "g.E().hasLabel(\"knows\").has(\"weight\",1)"
+    And the graph should return 1 for count of "g.E().hasLabel(\"knows\").has(\"weight\",0)"
+    And the graph should return 0 for count of "g.V().hasLabel(\"knows\").has(\"weight\")"
```
Comment (vkagamlyk) [inline gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/MergeEdge.feature:885]:
```suggestion
    And the graph should return 0 for count of "g.V().has(\"weight\")"
```

File: gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/MergeEdge.feature (hunk 1 above)
Comment (spmallette) [reply gremlin-test/src/main/resources/org/apache/tinkerpop/gremlin/test/features/map/MergeEdge.feature]: bad cut/paste - fixed both - thanks.

**Comment by vkagamlyk** [comment]
VOTE +1

not exactly related to the current ticket, but is there same issue for `MergeV`?

**Comment by xiazcy** [comment]
VOTE +1

**Comment by Cole-Greer** [comment]
LGTM VOTE +1

**Pull request merged**